src/merkle/simple_smt/mod.rs

```diff
@@ -0,0 +1,249 @@
+use super::{BTreeMap, MerkleError, Rpo256, RpoDigest, Vec, Word};
+
+#[cfg(test)]
+mod tests;
+
+// SPARSE MERKLE TREE
+// ================================================================================================
+
+/// A sparse Merkle tree with 63-bit keys and 256-bit leaf values, and with no compaction. Manipulation
+/// and retrieval of leaves and internal nodes is provided by the Store. The root of the tree is
+/// recomputed on each new leaf update.
+///
+/// This struct is intended to be used as one of the variants of the MerkleSet enum.
+#[derive(Clone, Debug)]
+pub struct SimpleSmt {
+    root: Word,
+    depth: u32,
+    store: Store,
+}
+
+impl SimpleSmt {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// Minimum supported depth.
+    pub const MIN_DEPTH: u32 = 1;
+
+    /// Maximum supported depth.
+    pub const MAX_DEPTH: u32 = 63;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn new(keys: Vec<u64>, values: Vec<Word>, depth: u32) -> Result<Self, MerkleError> {
+        // validate the range of the depth.
+        if depth < Self::MIN_DEPTH {
+            return Err(MerkleError::DepthTooSmall(depth));
+        } else if Self::MAX_DEPTH < depth {
+            return Err(MerkleError::DepthTooBig(depth));
+        }
+
```

Comment:
@bobbinth do you think it's good to preserve this API too? IMO it would be better like this since it will be guaranteed to be sound, out of the box

```rust
pub fn novak<I>(entries: I, depth: u32) -> Result<Self, MerkleError>
where
    I: Iterator<Item = (u64, Word)> + ExactSizeIterator,
```

Reply:
Agreed! Let's use the API you've suggested.